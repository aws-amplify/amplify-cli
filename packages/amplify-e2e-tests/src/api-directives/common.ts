import path from 'path';
import fs from 'fs-extra';
import _ from 'lodash';
import gql from 'graphql-tag';
import { addApi, amplifyPush, updateAuthAddUserGroups } from 'amplify-e2e-core';

import {
  setupUser,
  getUserPoolId,
  getApiKey,
  configureAmplify,
  signInUser,
  getConfiguredAppsyncClientAPIKeyAuth,
  getConfiguredAppsyncClientCognitoAuth,
} from './authHelper';

const GROUPNAME = 'Admin';
const USERNAME = 'user1';
const PASSWORD = 'user1Password';

export async function runTest(projectDir: string, testModule: any) {
  await addApi(projectDir);
  updateSchemaInTestProject(projectDir, testModule.schema);
  await amplifyPush(projectDir);

  const awsconfig = configureAmplify(projectDir);
  const apiKey = getApiKey(projectDir);
  const appSyncClient = getConfiguredAppsyncClientAPIKeyAuth(awsconfig.aws_appsync_graphqlEndpoint, awsconfig.aws_appsync_region, apiKey);

  await testMutations(testModule, appSyncClient);
  await testQueries(testModule, appSyncClient);
}

export async function runAutTest(projectDir: string, testModule: any) {
  await addApi(projectDir, {
    'Amazon Cognito User Pool': {},
  });
  updateSchemaInTestProject(projectDir, testModule.schema);

  await updateAuthAddUserGroups(projectDir, [GROUPNAME]);
  await amplifyPush(projectDir);
  const awsconfig = configureAmplify(projectDir);

  const userPoolId = getUserPoolId(projectDir);
  await setupUser(userPoolId, USERNAME, PASSWORD, GROUPNAME);

  const user = await signInUser(USERNAME, PASSWORD);
  const appSyncClient = getConfiguredAppsyncClientCognitoAuth(awsconfig.aws_appsync_graphqlEndpoint, awsconfig.aws_appsync_region, user);

  await testMutations(testModule, appSyncClient);
  await testQueries(testModule, appSyncClient);
  await testSubscriptions(testModule, appSyncClient);
}

export async function runMultiAutTest(projectDir: string, testModule: any) {
  await addApi(projectDir, {
    'API key': {},
    'Amazon Cognito User Pool': {},
    IAM: {},
  });
  updateSchemaInTestProject(projectDir, testModule.schema);

  await updateAuthAddUserGroups(projectDir, [GROUPNAME]);
  await amplifyPush(projectDir);
  const awsconfig = configureAmplify(projectDir);

  const userPoolId = getUserPoolId(projectDir);
  await setupUser(userPoolId, USERNAME, PASSWORD, GROUPNAME);

  const user = await signInUser(USERNAME, PASSWORD);
  const appSyncClient = getConfiguredAppsyncClientCognitoAuth(awsconfig.aws_appsync_graphqlEndpoint, awsconfig.aws_appsync_region, user);

  await testMutations(testModule, appSyncClient);
  await testQueries(testModule, appSyncClient);
  await testSubscriptions(testModule, appSyncClient);
}

export function updateSchemaInTestProject(projectDir: string, schema: any) {
  const backendApiDirPath = path.join(projectDir, 'amplify', 'backend', 'api');
  const apiResDirName = fs.readdirSync(backendApiDirPath)[0];
  const backendSchemaFilePath = path.join(backendApiDirPath, apiResDirName, 'schema.graphql');
  fs.writeFileSync(backendSchemaFilePath, schema);
}

export async function testMutations(testModule: any, appSyncClient: any) {
  let mutationNames = Object.keys(testModule).filter(key => /^mutation[0-9]*$/.test(key));

  if (mutationNames.length > 1) {
    mutationNames = mutationNames.sort((name1, name2) => {
      const n1 = parseInt(name1.replace(/mutation/, ''));
      const n2 = parseInt(name2.replace(/mutation/, ''));
      return n1 - n2;
    });
  }

  const mutationTasks = [];
  mutationNames.forEach(mutationName => {
    const mutation = testModule[mutationName];
    const mutationInput = testModule['input_' + mutationName];
    const mutationResult = testModule['expected_result_' + mutationName];

    mutationTasks.push(async () => {
      await testMutation(appSyncClient, mutation, mutationInput, mutationResult);
    });
  });

  await runInSequential(mutationTasks);
}

export async function testMutation(appSyncClient: any, mutation: any, mutationInput?: any, mutationResult?: any) {
  let resultMatch = true;
  let errorMatch = true;
  let actualResponse;

  try {
    const result = await appSyncClient.mutate({
      mutation: gql(mutation),
      fetchPolicy: 'no-cache',
      variables: mutationInput,
    });
    if (!checkResult(result, mutationResult)) {
      actualResponse = result;
      resultMatch = false;
    }
  } catch (err) {
    if (!checkError(err, mutationResult)) {
      actualResponse = err;
      errorMatch = false;
    }
  }
  if (!resultMatch || !errorMatch) {
    console.log('The following mutation test failed.');
    console.log('Mutation: ', mutation);
    if (mutationInput) {
      console.log('Mutation input: ', mutationInput);
    }
    if (mutationResult) {
      console.log('Expected mutation result: ', mutationResult);
    }
    if (actualResponse) {
      console.log('Actual mutation response: ', actualResponse);
    }
    throw new Error('Mutation test failed.');
  }
}

export async function testQueries(testModule: any, appSyncClient: any) {
  let queryNames = Object.keys(testModule).filter(key => /^query[0-9]*$/.test(key));

  if (queryNames.length > 1) {
    queryNames = queryNames.sort((name1, name2) => {
      const n1 = parseInt(name1.replace(/query/, ''));
      const n2 = parseInt(name2.replace(/query/, ''));
      return n1 - n2;
    });
  }

  const queryTasks = [];
  queryNames.forEach(queryName => {
    const query = testModule[queryName];
    const queryInput = testModule['input_' + queryName];
    const queryResult = testModule['expected_result_' + queryName];

    queryTasks.push(async () => {
      await testQuery(appSyncClient, query, queryInput, queryResult);
    });
  });

  await runInSequential(queryTasks);
}

export async function testQuery(appSyncClient: any, query: any, queryInput?: any, queryResult?: any) {
  let resultMatch = true;
  let errorMatch = true;
  let actualResponse;
  try {
    const result = await appSyncClient.query({
      query: gql(query),
      fetchPolicy: 'no-cache',
      variables: queryInput,
    });
    if (!checkResult(result, queryResult)) {
      actualResponse = result;
      resultMatch = false;
    }
  } catch (err) {
    if (!checkError(err, queryResult)) {
      actualResponse = err;
      errorMatch = false;
    }
  }
  if (!resultMatch || !errorMatch) {
    console.log('The following query test failed.');
    console.log('Query: ', query);
    if (queryInput) {
      console.log('Query input: ', queryInput);
    }
    if (queryResult) {
      console.log('Expected query result: ', queryResult);
    }
    if (actualResponse) {
      console.log('Actual query response: ', actualResponse);
    }
    throw new Error('Query test failed.');
  }
}

export async function testSubscriptions(testModule: any, appsyncClient: any) {
  let subscriptionNames = Object.keys(testModule).filter(key => /^subscription[0-9]*$/.test(key));

  if (subscriptionNames.length > 1) {
    subscriptionNames = subscriptionNames.sort((name1, name2) => {
      const n1 = parseInt(name1.replace(/subscription/, ''));
      const n2 = parseInt(name2.replace(/subscription/, ''));
      return n1 - n2;
    });
  }

  const subscriptionTasks = [];
  subscriptionNames.forEach(subscriptionName => {
    const subscription = testModule[subscriptionName];
    const subscriptionInput = testModule['input_' + subscriptionName];
    const subscriptionResult = testModule['expected_result_' + subscriptionName];
    const mutations = testModule['mutations_' + subscriptionName];
    const mutationsInput = testModule['input_mutations_' + subscriptionName];

    subscriptionTasks.push(async () => {
      await testSubscription(appsyncClient, subscription, mutations, subscriptionResult, subscriptionInput, mutationsInput);
    });
  });

  await runInSequential(subscriptionTasks);
}

export async function testSubscription(
  appSyncClient: any,
  subscription: string,
  mutations: any[],
  subscriptionResult: any,
  subscriptionInput?: any,
  mutationInputs?: any[],
) {
  const observer = appSyncClient.subscribe({
    query: gql(subscription),
    variables: subscriptionInput,
  });

  const received = [];
  const sub = observer.subscribe((event: any) => {
    received.push(event.data);
  });

  await new Promise(res => setTimeout(() => res(), 4000));

  const mutationTasks = [];
  for (let i = 0; i < mutations.length; i++) {
    const mutation = mutations[i];
    const mutationInput = mutationInputs ? mutationInputs[i] : undefined;
    mutationTasks.push(async () => {
      await appSyncClient.mutate({
        mutation: gql(mutation),
        fetchPolicy: 'no-cache',
        variables: mutationInput,
      });
      await new Promise(res => setTimeout(() => res(), 4000)); //to ensure correct order in received data
    });
  }

  await runInSequential(mutationTasks);

  await new Promise(res => setTimeout(() => res(), 4000));

  sub.unsubscribe();
  if (!checkResult(received, subscriptionResult)) {
    console.log('The following subscription test failed.');
    console.log('Subscription: ', subscription);
    if (subscriptionResult) {
      console.log('Expected subscription result: ', subscriptionResult);
    }
    if (received) {
      console.log('Actual subscription response: ', received);
    }
    throw new Error('Subscription test failed.');
  }
}

/////
function checkResult(received: any, expected: any): boolean {
  if (!expected) {
    //the test does not request result check, as long as the mutation/query goes through, it's good
    return true;
  }
  const queue = [
    {
      received,
      expected,
      depth: 0,
    },
  ];
  try {
    return runCompare(queue);
  } catch (e) {
    console.log('checkResult error: ', e);
    return false;
  }
}

function checkError(received: any, expected: any): boolean {
  if (!expected) {
    //the test does not request result check, assume mutation/query should go through, but received error
    return false;
  }
  const queue = [
    {
      received,
      expected,
      depth: 0,
    },
  ];
  return runCompare(queue);
}

const MAX_DEPTH = 50;
function runCompare(queue: { received: any; expected: any; depth: number }[]): boolean {
  let result = true;

  while (queue.length > 0 && result) {
    const itemToCompare = queue.shift();
    if (itemToCompare.depth > MAX_DEPTH) {
      break;
    }
    if (typeof itemToCompare.expected === 'object') {
      if (itemToCompare.expected === null) {
        result = itemToCompare.received === null;
      } else if (itemToCompare.received === null) {
        result = false;
      } else if (typeof itemToCompare.received === 'object') {
        Object.keys(itemToCompare.expected).forEach(key => {
          queue.push({
            received: itemToCompare.received[key],
            expected: itemToCompare.expected[key],
            depth: itemToCompare.depth + 1,
          });
        });
      } else {
        result = false;
      }
    } else if (itemToCompare.expected === '<check-defined>') {
      result = itemToCompare.received !== null && itemToCompare.received !== undefined;
    } else {
      result = itemToCompare.received === itemToCompare.expected;
    }
  }

  return result;
}

export async function runInSequential(tasks: ((v: any) => Promise<any>)[]): Promise<any> {
  let result;

  for (const task of tasks) {
    result = await task(result);
  }

  return result;
}
