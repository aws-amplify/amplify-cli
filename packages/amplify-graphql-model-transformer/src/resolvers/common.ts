import {
  CompoundExpressionNode,
  Expression,
  comment,
  iff,
  and,
  ref,
  notEquals,
  methodCall,
  int,
  compoundExpression,
  set,
  obj,
  equals,
  ifElse,
  printBlock,
  toJson,
  qref,
  str,
  not,
} from 'graphql-mapping-template';
import { OPERATION_KEY } from '../definitions';

const API_KEY = 'API Key Authorization';

/**
 * Helper method to generate code that converts DynamoDB condition object to condition
 * expression
 * @param inputConditionObjectName : Variable in stash that holds condition object
 * @param conditionOutputVariableName: Variable to store generated DDB expression
 */
export const generateConditionSlot = (inputConditionObjectName: string, conditionOutputVariableName: string): CompoundExpressionNode => {
  const statements: Expression[] = [
    comment('Start condition block'),

    iff(
      and([ref(inputConditionObjectName), notEquals(methodCall(ref(`${inputConditionObjectName}.size`)), int(0))]),
      compoundExpression([
        set(ref('mergedConditions'), obj({ and: ref(inputConditionObjectName) })),
        set(
          ref(conditionOutputVariableName),
          methodCall(ref('util.parseJson'), methodCall(ref('util.transform.toDynamoDBConditionExpression'), ref('mergedConditions'))),
        ),
        iff(
          and([
            ref(`${conditionOutputVariableName}.expressionValues`),
            equals(methodCall(ref(`${conditionOutputVariableName}.expressionValues.size`)), int(0)),
          ]),
          set(
            ref(conditionOutputVariableName),
            obj({
              expression: ref(`${conditionOutputVariableName}.expression`),
              expressionNames: ref(`${conditionOutputVariableName}.expressionNames`),
            }),
          ),
        ),
        comment('End condition block'),
      ]),
    ),
  ];
  return compoundExpression(statements);
};

/**
 * Generate common response template used by most of the resolvers.
 * Append operation if response is coming from a mutation, this is to protect field resolver for subscriptions
 */
export const generateDefaultResponseMappingTemplate = (isSyncEnabled: boolean, mutation = false): string => {
  const statements: Expression[] = [];
  if (mutation) statements.push(qref(methodCall(ref('ctx.result.put'), str(OPERATION_KEY), str('Mutation'))));
  if (isSyncEnabled) {
    statements.push(
      ifElse(
        ref('ctx.error'),
        methodCall(ref('util.error'), ref('ctx.error.message'), ref('ctx.error.type'), ref('ctx.result')),
        toJson(ref('ctx.result')),
      ),
    );
  } else {
    statements.push(
      ifElse(ref('ctx.error'), methodCall(ref('util.error'), ref('ctx.error.message'), ref('ctx.error.type')), toJson(ref('ctx.result'))),
    );
  }

  return printBlock('ResponseTemplate')(compoundExpression(statements));
};

/**
 * Util function to generate resolver key used to keep track of all the resolvers in memory
 * @param typeName Name of the type
 * @param fieldName Name of the field
 */
export const generateResolverKey = (typeName: string, fieldName: string): string => {
  return `${typeName}.${fieldName}`;
};

/**
 * Util function to generate sandbox mode expression
 */
export const generateAuthExpressionForSandboxMode = (enabled: boolean): string => {
  let exp;

  if (enabled) exp = iff(notEquals(methodCall(ref('util.authType')), str(API_KEY)), methodCall(ref('util.unauthorized')));
  else exp = methodCall(ref('util.unauthorized'));

  return printBlock(`Sandbox Mode ${enabled ? 'Enabled' : 'Disabled'}`)(
    compoundExpression([iff(not(ref('ctx.stash.get("hasAuth")')), exp), toJson(obj({}))]),
  );
};
