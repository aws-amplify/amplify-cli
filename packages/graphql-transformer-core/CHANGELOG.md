# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="1.0.28"></a>
## [1.0.28](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.28-beta.0...graphql-transformer-core@1.0.28) (2018-10-18)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.28-beta.0"></a>
## [1.0.28-beta.0](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.12...graphql-transformer-core@1.0.28-beta.0) (2018-10-12)


### Bug Fixes

* **graphql-transformer-core:** Fix Readme.md docs for the `[@auth](https://github.com/auth)` directive ([db6ff7a](https://github.com/aws-amplify/amplify-cli/commit/db6ff7a))




<a name="1.0.12"></a>
## [1.0.12](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.11...graphql-transformer-core@1.0.12) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.11"></a>
## [1.0.11](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.10...graphql-transformer-core@1.0.11) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.10"></a>
## [1.0.10](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.9...graphql-transformer-core@1.0.10) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.9"></a>
## [1.0.9](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.8...graphql-transformer-core@1.0.9) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.8"></a>
## [1.0.8](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.7...graphql-transformer-core@1.0.8) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.7"></a>
## [1.0.7](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.6...graphql-transformer-core@1.0.7) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.6"></a>
## [1.0.6](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.5...graphql-transformer-core@1.0.6) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.5"></a>
## [1.0.5](https://github.com/aws-amplify/amplify-cli/compare/graphql-transformer-core@1.0.4...graphql-transformer-core@1.0.5) (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.4"></a>
## 1.0.4 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.3"></a>
## 1.0.3 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.2"></a>
## 1.0.2 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core

<a name="1.0.1"></a>
## 1.0.1 (2018-08-23)




**Note:** Version bump only for package graphql-transformer-core
