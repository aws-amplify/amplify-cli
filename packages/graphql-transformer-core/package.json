{
  "name": "graphql-transformer-core",
  "version": "6.19.4",
  "description": "A framework to transform from GraphQL SDL to AWS CloudFormation.",
  "repository": {
    "type": "git",
    "url": "https://github.com/aws-amplify/amplify-cli.git",
    "directory": "packages/graphql-transformer-core"
  },
  "author": "Amazon Web Services",
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "keywords": [
    "graphql",
    "cloudformation",
    "aws"
  ],
  "scripts": {
    "test": "jest",
    "test-ci": "jest --ci -i",
    "build": "tsc",
    "watch": "tsc -w",
    "clean": "rimraf ./lib"
  },
  "dependencies": {
    "cloudform-types": "^4.2.0",
    "deep-diff": "^1.0.2",
    "fs-extra": "^8.1.0",
    "glob": "^7.1.6",
    "graphql": "^14.5.8",
    "graphql-transformer-common": "4.17.4"
  },
  "devDependencies": {
    "@types/fs-extra": "^8.0.1",
    "@types/node": "^10.17.13"
  },
  "jest": {
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    },
    "testURL": "http://localhost",
    "testRegex": "(src/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$",
    "moduleFileExtensions": [
      "ts",
      "tsx",
      "js",
      "jsx",
      "json",
      "node"
    ],
    "collectCoverage": true
  }
}
