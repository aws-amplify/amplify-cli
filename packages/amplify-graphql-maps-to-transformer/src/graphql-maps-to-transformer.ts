import { TransformerPluginBase, InvalidDirectiveError } from '@aws-amplify/graphql-transformer-core';
import {
  TransformerContextProvider,
  TransformerPluginType,
  TransformerSchemaVisitStepContextProvider,
} from '@aws-amplify/graphql-transformer-interfaces';
import { ObjectTypeDefinitionNode, DirectiveNode, Kind, DefinitionNode } from 'graphql';
import { createMappingLambda } from './field-mapping-lambda';
import { attachFilterAndConditionInputMappingSlot, attachInputMappingSlot, attachResponseMappingSlot } from './field-mapping-resolvers';

const directiveName = 'mapsTo';

const directiveDefinition = `
  directive @${directiveName}(name: String!) on OBJECT
`;

export class MapsToTransformer extends TransformerPluginBase {
  constructor() {
    super(`amplify-maps-to-transformer`, directiveDefinition, TransformerPluginType.GENERIC);
  }

  /**
   * During the AST tree walking, the mapsTo transformer registers any renamed models with the ctx.resourceHelper
   */
  object = (definition: ObjectTypeDefinitionNode, directive: DirectiveNode, ctx: TransformerSchemaVisitStepContextProvider) => {
    const modelName = definition.name.value;
    const prevNameNode = directive.arguments?.find(arg => arg.name.value === 'name');

    const hasModelDirective = !!definition.directives?.find(directive => directive.name.value === 'model');
    if (!hasModelDirective) {
      throw new InvalidDirectiveError(`@mapsTo can only be used on an @model type`);
    }

    // the following checks should never fail because the graphql schema already validates them, but TS complains without them
    if (!prevNameNode) {
      throw new InvalidDirectiveError(`name is required in @${directiveName} directive`);
    }

    if (prevNameNode.value.kind !== 'StringValue') {
      throw new InvalidDirectiveError(`A single string must be provided for "name" in @${directiveName} directive`);
    }

    const originalName = prevNameNode.value.value;

    const schemaHasConflictingModel = !!ctx.inputDocument.definitions.find(hasModelWithNamePredicate(originalName));
    if (schemaHasConflictingModel) {
      throw new InvalidDirectiveError(`Type ${modelName} cannot map to ${originalName} because ${originalName} is a model in the schema.`);
    }
    ctx.resourceHelper.setModelNameMapping(modelName, originalName);
  };

  /**
   * During the generateResolvers step, the mapsTo transformer reads all of the model field mappings from the resourceHelper and generates
   * VTL to map the current field names to the original field names
   */
  after = (context: TransformerContextProvider) => {
    context.resourceHelper.getModelFieldMapKeys().forEach(modelName => {
      const modelFieldMap = context.resourceHelper.getModelFieldMap(modelName);
      if (!modelFieldMap.getMappedFields().length) {
        return;
      }
      const lambdaDataSource = createMappingLambda(context.api.host, context.stackManager);
      modelFieldMap.getResolverReferences().forEach(({ typeName, fieldName, isList }) => {
        const resolver = context.resolvers.getResolver(typeName, fieldName);
        if (!resolver) {
          return;
        }
        if (typeName === 'Mutation') {
          attachInputMappingSlot({
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
          });
          attachFilterAndConditionInputMappingSlot({
            slotName: 'preUpdate',
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
            dataSource: lambdaDataSource,
          });
          attachResponseMappingSlot({
            slotName: 'postUpdate',
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
            isList: false,
          });
        } else {
          attachFilterAndConditionInputMappingSlot({
            slotName: 'preDataLoad',
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
            dataSource: lambdaDataSource,
          });
          attachResponseMappingSlot({
            slotName: 'postDataLoad',
            resolver,
            resolverTypeName: typeName,
            resolverFieldName: fieldName,
            fieldMap: modelFieldMap.getMappedFields(),
            isList,
          });
        }
      });
    });
  };
}

// returns a predicate for determining if a DefinitionNode is an model object with the given name
const hasModelWithNamePredicate = (name: string) => (node: DefinitionNode) =>
  node.kind === Kind.OBJECT_TYPE_DEFINITION && !!node.directives?.find(dir => dir.name.value === 'model') && node.name.value === name;
