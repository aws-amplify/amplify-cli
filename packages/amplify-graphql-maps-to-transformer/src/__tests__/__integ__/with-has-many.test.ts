import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { GraphQLTransform } from '@aws-amplify/graphql-transformer-core';
import { BelongsToTransformer, HasManyTransformer } from '@aws-amplify/graphql-relational-transformer';
import { MapsToTransformer } from '../../graphql-maps-to-transformer';
import { expectedResolversForModelWithRenamedField } from './common';

const mappedHasMany = /* GraphQL */ `
  type Employee @model @mapsTo(name: "Person") {
    id: ID!
    tasks: [Task] @hasMany
  }

  type Task @model {
    id: ID!
    title: String
  }
`;

const transformSchema = (schema: string) => {
  const transformer = new GraphQLTransform({
    transformers: [new ModelTransformer(), new HasManyTransformer(), new BelongsToTransformer(), new MapsToTransformer()],
    sandboxModeEnabled: true,
  });
  return transformer.transform(schema);
};

describe('@mapsTo with @hasMany', () => {
  it('adds CRUD input and output mappings on related type and maps related type in hasMany field resolver', () => {
    const out = transformSchema(mappedHasMany);
    const expectedResolvers: string[] = expectedResolversForModelWithRenamedField('Task').concat('Employee.tasks.postDataLoad.1.res.vtl');
    expectedResolvers.forEach(resolver => {
      expect(out.resolvers[resolver]).toMatchSnapshot();
    });
  });
});
