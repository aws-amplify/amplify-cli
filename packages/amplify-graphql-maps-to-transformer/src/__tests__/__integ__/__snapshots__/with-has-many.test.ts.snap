// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 1`] = `
"$util.qr($ctx.args.input.put(\\"personTasksId\\", $ctx.args.input.employeeTasksId))
$util.qr($ctx.args.input.remove(\\"employeeTasksId\\"))
$util.toJson({})"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 2`] = `
"#set( $fieldMap = {\\"employeeTasksId\\":\\"personTasksId\\"} )
#if( $util.isNull($fieldMap) || $fieldMap.keySet().size() <= 0 )
  #return($ctx.args)
#end
#if( $util.isNull($ctx.args.filter) && $util.isNull($ctx.args.condition) && $util.isNull($ctx.args.sort) && $util.isNull($ctx.args.aggregates) )
  #return($ctx.args)
#end
#set( $invoke = {
  \\"operation\\": \\"Invoke\\",
  \\"payload\\": {
      \\"args\\": $ctx.args,
      \\"fieldMap\\": $fieldMap
  }
} )
$util.toJson($invoke)"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 3`] = `
"#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
#set( $ctx.stash.transformedArgs = $ctx.result )
$util.toJson({})"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 4`] = `
"$util.qr($ctx.prev.result.put(\\"employeeTasksId\\", $ctx.prev.result.personTasksId))
$util.qr($ctx.prev.result.remove(\\"personTasksId\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 5`] = `
"$util.qr($ctx.args.input.put(\\"personTasksId\\", $ctx.args.input.employeeTasksId))
$util.qr($ctx.args.input.remove(\\"employeeTasksId\\"))
$util.toJson({})"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 6`] = `
"#set( $fieldMap = {\\"employeeTasksId\\":\\"personTasksId\\"} )
#if( $util.isNull($fieldMap) || $fieldMap.keySet().size() <= 0 )
  #return($ctx.args)
#end
#if( $util.isNull($ctx.args.filter) && $util.isNull($ctx.args.condition) && $util.isNull($ctx.args.sort) && $util.isNull($ctx.args.aggregates) )
  #return($ctx.args)
#end
#set( $invoke = {
  \\"operation\\": \\"Invoke\\",
  \\"payload\\": {
      \\"args\\": $ctx.args,
      \\"fieldMap\\": $fieldMap
  }
} )
$util.toJson($invoke)"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 7`] = `
"#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
#set( $ctx.stash.transformedArgs = $ctx.result )
$util.toJson({})"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 8`] = `
"$util.qr($ctx.prev.result.put(\\"employeeTasksId\\", $ctx.prev.result.personTasksId))
$util.qr($ctx.prev.result.remove(\\"personTasksId\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 9`] = `
"$util.qr($ctx.args.input.put(\\"personTasksId\\", $ctx.args.input.employeeTasksId))
$util.qr($ctx.args.input.remove(\\"employeeTasksId\\"))
$util.toJson({})"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 10`] = `
"#set( $fieldMap = {\\"employeeTasksId\\":\\"personTasksId\\"} )
#if( $util.isNull($fieldMap) || $fieldMap.keySet().size() <= 0 )
  #return($ctx.args)
#end
#if( $util.isNull($ctx.args.filter) && $util.isNull($ctx.args.condition) && $util.isNull($ctx.args.sort) && $util.isNull($ctx.args.aggregates) )
  #return($ctx.args)
#end
#set( $invoke = {
  \\"operation\\": \\"Invoke\\",
  \\"payload\\": {
      \\"args\\": $ctx.args,
      \\"fieldMap\\": $fieldMap
  }
} )
$util.toJson($invoke)"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 11`] = `
"#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
#set( $ctx.stash.transformedArgs = $ctx.result )
$util.toJson({})"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 12`] = `
"$util.qr($ctx.prev.result.put(\\"employeeTasksId\\", $ctx.prev.result.personTasksId))
$util.qr($ctx.prev.result.remove(\\"personTasksId\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 13`] = `
"$util.qr($ctx.prev.result.put(\\"employeeTasksId\\", $ctx.prev.result.personTasksId))
$util.qr($ctx.prev.result.remove(\\"personTasksId\\"))
$util.toJson($ctx.prev.result)"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 14`] = `
"#set( $fieldMap = {\\"employeeTasksId\\":\\"personTasksId\\"} )
#if( $util.isNull($fieldMap) || $fieldMap.keySet().size() <= 0 )
  #return($ctx.args)
#end
#if( $util.isNull($ctx.args.filter) && $util.isNull($ctx.args.condition) && $util.isNull($ctx.args.sort) && $util.isNull($ctx.args.aggregates) )
  #return($ctx.args)
#end
#set( $invoke = {
  \\"operation\\": \\"Invoke\\",
  \\"payload\\": {
      \\"args\\": $ctx.args,
      \\"fieldMap\\": $fieldMap
  }
} )
$util.toJson($invoke)"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 15`] = `
"#if( $ctx.error )
  $util.error($ctx.error.message, $ctx.error.type)
#end
#set( $ctx.stash.transformedArgs = $ctx.result )
$util.toJson({})"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 16`] = `
"#foreach( $item in $ctx.prev.result.items )
  $util.qr($item.put(\\"employeeTasksId\\", $item.personTasksId))
  $util.qr($item.remove(\\"personTasksId\\"))
#end
$util.toJson($ctx.prev.result)"
`;

exports[`@mapsTo with @hasMany adds CRUD input and output mappings on related type and maps related type in hasMany field resolver 17`] = `
"#foreach( $item in $ctx.prev.result.items )
  $util.qr($item.put(\\"employeeTasksId\\", $item.personTasksId))
  $util.qr($item.remove(\\"personTasksId\\"))
#end
$util.toJson($ctx.prev.result)"
`;
